src/main/scala/net/snowflake/spark/snowflake/pushdowns/SnowflakeScanExec.scala

```diff
@@ -0,0 +1,75 @@
+package net.snowflake.spark.snowflake.pushdowns
+
+import net.snowflake.spark.snowflake.{SnowflakeRelation, SnowflakeSQLStatement}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, UnsafeProjection}
+import org.apache.spark.sql.execution.LeafExecNode
+
+import java.util.concurrent.{Callable, ExecutorService, Executors, Future}
+
+/**
+ * Snowflake Scan Plan for pushing query fragment to snowflake endpoint
+ *
+ * @param projection   projected columns
+ * @param snowflakeSQL SQL query that is pushed to snowflake for evaluation
+ * @param relation     Snowflake datasource
+ */
+case class SnowflakeScanExec(projection: Seq[Attribute],
+                             snowflakeSQL: SnowflakeSQLStatement,
+                             relation: SnowflakeRelation) extends LeafExecNode {
+  // result holder
+  @transient implicit private var data: Future[PushDownResult] = _
+  @transient implicit private val service: ExecutorService = Executors.newCachedThreadPool()
+
+  override protected def doPrepare(): Unit = {
```

Comment:
What is benefit of building RDD in doPrepare instead of doExecute?

Reply:
The doPrepare method allows spark planner in performing initial metadata collection work in async fashion. While do execute is always blocking call. Thus leveraging doPrepare we can move some work such as building sql and creating connection with snowflake in background thread while main planner operates on other nodes in the plan giving some perf gains